master/static/srv/ship_logs.py

```diff
@@ -0,0 +1,598 @@
+    token: str,
+    emit_stdout_logs: bool,
+    cmd: List[str],
+    log_wait_time: int,
+) -> int:
+    logq = queue.Queue()  # type: queue.Queue
+    doneq = queue.Queue()  # type: queue.Queue
+
+    waiter_started = False
+    stdout_started = False
+    stderr_started = False
+    shipper_started = False
+
+    # Normally we like structured concurrency; i.e. a function that owns a thread must not exit
+    # until that thead has been properly cleaned up.  However, it is important that the log shipper
+    # is not allowed to keep a task container alive too long after the child process has exited.  We
+    # want to guarantee that we exit about DET_LOG_WAIT_TIME seconds after the child process exits.
+    #
+    # However, interruping a synchronous HTTP call form urllib is nearly impossible; even if you
+    # were to select() until the underlying file descriptor had something to read before calling
+    # Request.read(), there are many buffered readers in there and most likely multiple os.read()
+    # calls would occur and you'd be blocking anyway.
+    #
+    # So as an easy workaround, we set daemon=True and just exit the process if it's not done on
+    # time.
+    shipper = Shipper(master_url, token, cert_name, cert_file, logq, doneq, daemon=True)
+    shipper_timed_out = False
+
+    # Start the process or ship a special log message to the master why we couldn't.
+    try:
+        # Don't rely on Popen's standard line buffering; we want to do our own line buffering.
+        p = subprocess.Popen(
+            cmd, stdin=subprocess.DEVNULL, stdout=subprocess.PIPE, stderr=subprocess.PIPE, bufsize=0
+        )
+    except FileNotFoundError:
+        shipper.ship_special(f"FileNotFoundError executing {cmd}", metadata, emit_stdout_logs)
+        # 127 is the standard bash exit code for file-not-found.
+        return 127
+    except PermissionError:
+        # Unable to read or to execute the command.
+        shipper.ship_special(f"PermissionError executing {cmd}", metadata, emit_stdout_logs)
+        # 126 is the standard bash exit code for permission failure.
+        return 126
+    except Exception:
+        msg = f"unexpected failure executing {cmd}:\n" + traceback.format_exc()
+        shipper.ship_special(msg, metadata, emit_stdout_logs)
+        # 80 is the exit code we use to signal "ship_logs.py failed"
+        return 80
+
+    # Just for mypy.
+    assert isinstance(p.stdout, io.RawIOBase) and isinstance(p.stderr, io.RawIOBase)
+
+    try:
+        stdout = Collector(p.stdout, "stdout", emit_stdout_logs, metadata, logq, doneq)
+        stderr = Collector(p.stderr, "stderr", emit_stdout_logs, metadata, logq, doneq)
+        waiter = threading.Thread(target=lambda: doneq.put(("waiter", p.wait(), None)))
+
+        waiter.start()
+        waiter_started = True
+
+        # Set up signal forwarding.
+        def signal_passthru(signum: Any, frame: Any):
+            p.send_signal(signum)
+
+        for sig in [
```

Comment:
This is nice. I wouldn't have thought to do it, but it looks to me like the right thing to do.